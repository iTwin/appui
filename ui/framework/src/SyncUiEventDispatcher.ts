/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */

import { UiEvent } from "@bentley/ui-core";

/** SyncUi Event arguments. Contains a set of lower case event Ids.
 */
export interface SyncUiEventArgs {
  eventIds: Set<string>;
}

/** SyncUi Event class.
 */
export class SyncUiEvent extends UiEvent<SyncUiEventArgs> { }

/** SyncUi Event Dispatcher class. This class is used to send eventIds to interested Ui components so the component can determine if it needs
 * to refresh its display by calling setState on itself.
 */
export class SyncUiEventDispatcher {
  private static _syncEventTimer: NodeJS.Timer | undefined;
  private static _eventIds: Set<string>;
  private static _eventIdAdded: boolean = false;
  private static _syncUiEvent: SyncUiEvent;
  private static _timeoutPeriod: number = 500;

  /** @hidden - used for testing only */
  public static setTimeoutPeriod(period: number): void {
    SyncUiEventDispatcher._timeoutPeriod = period;
  }

  /** Return SyncUiEvent so callers can register an event callback. */
  public static get syncEventIds(): Set<string> {
    if (!SyncUiEventDispatcher._eventIds)
      SyncUiEventDispatcher._eventIds = new Set<string>();

    return SyncUiEventDispatcher._eventIds;
  }

  /** Return SyncUiEvent so callers can register an event callback. */
  public static get onSyncUiEvent(): SyncUiEvent {
    if (!SyncUiEventDispatcher._syncUiEvent)
      SyncUiEventDispatcher._syncUiEvent = new SyncUiEvent();

    return SyncUiEventDispatcher._syncUiEvent;
  }

  /** Immediately trigger sync event processing. */
  public static dispatchImmediateSyncUiEvent(eventId: string): void {
    const eventIds = new Set<string>();
    eventIds.add(eventId.toLowerCase());
    SyncUiEventDispatcher.onSyncUiEvent.emit({ eventIds });
  }

  /** Save eventId in Set for processing. */
  public static dispatchSyncUiEvent(eventId: string): void {
    SyncUiEventDispatcher.syncEventIds.add(eventId.toLowerCase());
    SyncUiEventDispatcher._eventIdAdded = true;
    if (!SyncUiEventDispatcher._syncEventTimer) {  // if there is not a timer active, create one
      SyncUiEventDispatcher._syncEventTimer = setTimeout(SyncUiEventDispatcher.checkForAdditionalIds, SyncUiEventDispatcher._timeoutPeriod);
    }
  }

  /** Save multiple eventIds in Set for processing. */
  public static dispatchSyncUiEvents(eventIds: string[]): void {
    eventIds.forEach((id) => SyncUiEventDispatcher.syncEventIds.add(id.toLowerCase()));
    if (!SyncUiEventDispatcher._syncEventTimer) {  // if there is not a timer active, create one
      SyncUiEventDispatcher._syncEventTimer = setTimeout(SyncUiEventDispatcher.checkForAdditionalIds, SyncUiEventDispatcher._timeoutPeriod);
    } else {
      SyncUiEventDispatcher._eventIdAdded = true;
    }
  }

  /** Trigger registered event processing when timer has expired and no addition eventId are added. */
  private static checkForAdditionalIds() {
    if (!SyncUiEventDispatcher._eventIdAdded && SyncUiEventDispatcher._syncEventTimer) {
      if (SyncUiEventDispatcher._syncEventTimer) clearTimeout(SyncUiEventDispatcher._syncEventTimer);
      SyncUiEventDispatcher._syncEventTimer = undefined;
      SyncUiEventDispatcher._eventIdAdded = false;
      SyncUiEventDispatcher.onSyncUiEvent.emit({ eventIds: SyncUiEventDispatcher.syncEventIds });
      return;
    }

    if (SyncUiEventDispatcher._syncEventTimer) clearTimeout(SyncUiEventDispatcher._syncEventTimer);
    SyncUiEventDispatcher._eventIdAdded = false;
    SyncUiEventDispatcher._syncEventTimer = setTimeout(SyncUiEventDispatcher.checkForAdditionalIds, SyncUiEventDispatcher._timeoutPeriod);
  }
}
