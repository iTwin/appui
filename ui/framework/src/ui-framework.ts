/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

// cSpell:ignore safearea cursormenu clientservices oidc Textbox Modeless configurableui stagepanels dragdrop uiadmin

export * from "./ui-framework/UiFramework";  // Please ensure that this line comes before all other exports.

export * from "./ui-framework/CoreToolDefinitions";
export * from "./ui-framework/MarkupToolDefinitions";

export * from "./ui-framework/redux/SessionState";
export * from "./ui-framework/redux/FrameworkState";
export * from "./ui-framework/redux/connectIModel";

export * from "./ui-framework/clientservices/IModelServices";
export * from "./ui-framework/clientservices/ProjectServices";

export * from "./ui-framework/feedback/ValidationTextbox";
export * from "./ui-framework/feedback/ElementTooltip";

export * from "./ui-framework/messages/AppNotificationManager";
export * from "./ui-framework/messages/MessageManager";
export * from "./ui-framework/messages/InputField";
export * from "./ui-framework/messages/Pointer";

export * from "./ui-framework/oidc/SignIn";
export * from "./ui-framework/oidc/SignOut";

export * from "./ui-framework/imodel-components/spatial-tree/SpatialContainmentTree";
export * from "./ui-framework/imodel-components/category-tree/CategoriesTree";
export * from "./ui-framework/imodel-components/visibility-tree/VisibilityTree";

export * from "./ui-framework/pickers/ListPicker";
export * from "./ui-framework/pickers/ModelSelector/ModelSelector";
export * from "./ui-framework/pickers/ViewSelector";

export * from "./ui-framework/safearea/SafeAreaContext";

export * from "./ui-framework/configurableui/ConfigurableUiContent";
export * from "./ui-framework/configurableui/ConfigurableUiControl";
export * from "./ui-framework/configurableui/ConfigurableUiManager";
export * from "./ui-framework/configurableui/state";

export * from "./ui-framework/content/ContentControl";
export * from "./ui-framework/content/ContentGroup";
export * from "./ui-framework/content/ContentLayout";
export * from "./ui-framework/content/ContentLayoutProps";
export * from "./ui-framework/content/ContentLayoutManager";
export * from "./ui-framework/content/ContentViewManager";
export * from "./ui-framework/content/SavedView";
export * from "./ui-framework/content/SavedViewLayout";
export * from "./ui-framework/content/ViewportContentControl";
export * from "./ui-framework/content/IModelViewport";
export * from "./ui-framework/content/DefaultViewOverlay";

export * from "./ui-framework/dialog/DialogManagerBase";
export * from "./ui-framework/dialog/ModalDialogManager";
export * from "./ui-framework/dialog/ModelessDialog";
export * from "./ui-framework/dialog/ModelessDialogManager";
export * from "./ui-framework/dialog/StandardMessageBox";

export * from "./ui-framework/keyinbrowser/KeyinBrowser";

export * from "./ui-framework/dragdrop/DragDropLayerManager";
export * from "./ui-framework/dragdrop/ZoneTargets";

export * from "./ui-framework/frontstage/Frontstage";
export * from "./ui-framework/frontstage/FrontstageComposer";
export * from "./ui-framework/frontstage/FrontstageDef";
export * from "./ui-framework/frontstage/FrontstageManager";
export * from "./ui-framework/frontstage/FrontstageProvider";
export * from "./ui-framework/frontstage/NestedFrontstage";
export * from "./ui-framework/frontstage/ModalFrontstage";

export * from "./ui-framework/shared/ActionButtonItemDef";
export * from "./ui-framework/shared/AnyItemDef";
export * from "./ui-framework/shared/CommandItemDef";
export * from "./ui-framework/shared/ConditionalItemDef";
export * from "./ui-framework/shared/ConditionalItemProps";
export * from "./ui-framework/shared/CustomItemDef";
export * from "./ui-framework/shared/CustomItemProps";
export * from "./ui-framework/shared/GroupItemProps";
export * from "./ui-framework/shared/IconComponent";
export * from "./ui-framework/shared/ItemDefBase";
export * from "./ui-framework/shared/ItemMap";
export * from "./ui-framework/shared/ItemProps";
export * from "./ui-framework/shared/MenuItem";
export * from "./ui-framework/shared/ToolItemDef";

export * from "./ui-framework/keyboardshortcut/KeyboardShortcut";
export * from "./ui-framework/keyboardshortcut/KeyboardShortcutMenu";

export * from "./ui-framework/widgets/NavigationWidget";
export * from "./ui-framework/widgets/StatusBar";
export * from "./ui-framework/widgets/StatusBarWidgetControl";
export * from "./ui-framework/widgets/ToolbarWidgetBase";
export * from "./ui-framework/widgets/ToolWidget";
export * from "./ui-framework/widgets/Widget";
export * from "./ui-framework/widgets/WidgetControl";
export * from "./ui-framework/widgets/WidgetDef";
export * from "./ui-framework/widgets/WidgetHost";
export * from "./ui-framework/widgets/WidgetFactory";
export * from "./ui-framework/widgets/WidgetStack";
export * from "./ui-framework/widgets/realitydata/RealityDataPicker";
export * from "./ui-framework/widgets/VisibilityWidget";
export * from "./ui-framework/widgets/DefaultNavigationWidget";

export * from "./ui-framework/workflow/Task";
export * from "./ui-framework/workflow/Workflow";

export * from "./ui-framework/zones/FrameworkZone";
export * from "./ui-framework/zones/StatusBarZone";
export * from "./ui-framework/zones/toolsettings/ToolSettingsZone";
export * from "./ui-framework/zones/toolsettings/ToolInformation";
export * from "./ui-framework/zones/toolsettings/ToolUiManager";
export * from "./ui-framework/zones/toolsettings/ToolUiProvider";
export * from "./ui-framework/zones/toolsettings/DefaultToolSettingsProvider";

export * from "./ui-framework/tileloading/TileLoadingIndicator";

export * from "./ui-framework/zones/Zone";
export * from "./ui-framework/zones/ZoneDef";

export * from "./ui-framework/stagepanels/FrameworkStagePanel";
export * from "./ui-framework/stagepanels/StagePanel";
export * from "./ui-framework/stagepanels/StagePanelDef";
export * from "./ui-framework/stagepanels/StagePanelHeader";

export * from "./ui-framework/toolbar/ActionItemButton";
export * from "./ui-framework/toolbar/GroupItem";
export * from "./ui-framework/toolbar/PopupButton";
export * from "./ui-framework/toolbar/Toolbar";
export * from "./ui-framework/toolbar/ToolButton";

export * from "./ui-framework/cursor/CursorInformation";
export * from "./ui-framework/cursor/cursorprompt/CursorPrompt";
export * from "./ui-framework/cursor/cursorpopup/CursorPopup";
export * from "./ui-framework/cursor/cursorpopup/CursorPopupManager";
export * from "./ui-framework/cursor/cursormenu/CursorMenu";

export * from "./ui-framework/timeline/ScheduleAnimationProvider";
export * from "./ui-framework/timeline/AnalysisAnimationProvider";
export * from "./ui-framework/timeline/SolarTimelineDataProvider";

export * from "./ui-framework/backstage/Backstage";
export * from "./ui-framework/backstage/BackstageComposer";
export * from "./ui-framework/backstage/BackstageComposerItem";
export * from "./ui-framework/backstage/BackstageItem";
export * from "./ui-framework/backstage/BackstageItemsManager";
export * from "./ui-framework/backstage/BackstageItemUtilities";
export * from "./ui-framework/backstage/BackstageManager";
export * from "./ui-framework/backstage/CommandLaunch";
export * from "./ui-framework/backstage/FrontstageLaunch";
export * from "./ui-framework/backstage/Separator";
export * from "./ui-framework/backstage/TaskLaunch";
export * from "./ui-framework/backstage/UserProfile";

export * from "./ui-framework/navigationaids/NavigationAidControl";
export * from "./ui-framework/navigationaids/CubeNavigationAid";
export * from "./ui-framework/navigationaids/SheetNavigationAid";
export * from "./ui-framework/navigationaids/DrawingNavigationAid";
export * from "./ui-framework/navigationaids/SheetsModalFrontstage";
export * from "./ui-framework/navigationaids/StandardRotationNavigationAid";

export * from "./ui-framework/statusfields/ActivityCenter";
export * from "./ui-framework/statusfields/MessageCenter";
export * from "./ui-framework/statusfields/SnapMode";
export * from "./ui-framework/statusfields/PromptField";
export * from "./ui-framework/statusfields/SelectionInfo";
export * from "./ui-framework/statusfields/StatusFieldProps";
export * from "./ui-framework/statusfields/SelectionScope";
export * from "./ui-framework/statusfields/toolassistance/ToolAssistanceField";
export * from "./ui-framework/statusfields/Indicator";
export * from "./ui-framework/statusfields/SectionsField";
export * from "./ui-framework/statusfields/ViewAttributes";

export * from "./ui-framework/theme/ThemeManager";

export * from "./ui-framework/uiadmin/FrameworkUiAdmin";

export * from "./ui-framework/utils/ViewUtilities";
export * from "./ui-framework/utils/redux-ts";
export * from "./ui-framework/utils/PropsHelper";
export * from "./ui-framework/utils/UiShowHideManager";
export * from "./ui-framework/utils/ToolbarButtonHelper";

export * from "./ui-framework/syncui/SyncUiEventDispatcher";
export * from "./ui-framework/syncui/BooleanListener";

export * from "./ui-framework/accudraw/Calculator";
export * from "./ui-framework/accudraw/CalculatorEngine";
export * from "./ui-framework/accudraw/MenuButton";

export * from "./ui-framework/popup/PopupManager";
export * from "./ui-framework/popup/PositionPopup";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
// istanbul ignore next
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("ui-framework", BUILD_SEMVER);
}

/** @docs-package-description
 * The ui-framework package contains application fragments for Login, Project, iModel and View selection,
 * and configuration of the application UI with the Backstage, Frontstages, Widgets, etc.
 * For more information, see [learning about ui-framework]($docs/learning/ui/framework/index.md).
 */
/**
 * @docs-group-description Admin
 * APIs for various UI components, such as toolbars, buttons and menus.
 */
/**
 * @docs-group-description Backstage
 * Classes for working with a Backstage
 */
/**
 * @docs-group-description ConfigurableUi
 * Classes for working with the Application UI Configuration
 */
/**
 * @docs-group-description ContentView
 * Classes for working with a Content View, Group, Layout or Control
 */
/**
 * @docs-group-description Cursor
 * Cursor related information, components and events
 */
/**
 * @docs-group-description Dialog
 * Classes for working with a dialog
 */
/**
 * @docs-group-description DragDrop
 * Classes for managing DragDrop API drag layers
 */
/**
 * @docs-group-description Frontstage
 * Classes for working with a Frontstage
 */
/**
 * @docs-group-description IModelComponents
 * Classes for displaying information about an iModel
 */
/**
 * @docs-group-description Item
 * Classes for working with an Item in a Tool Widget, Navigation Widget or Backstage
 */
/**
 * @docs-group-description KeyboardShortcut
 * Classes for working with Keyboard Shortcuts
 */
/**
 * @docs-group-description NavigationAids
 * Classes for working with Navigation Aids
 */
/**
 * @docs-group-description Notification
 * Classes for working with a Notification or Message
 */
/**
 * @docs-group-description OIDC
 * Classes for working with the OpenID Connect (OIDC) protocol
 */
/**
 * @docs-group-description Picker
 * Classes for working with various pickers
 */
/**
 * @docs-group-description State
 * Classes for maintaining state
 */
/**
 * @docs-group-description StatusBar
 * Classes for defining a StatusBar
 */
/**
 * @docs-group-description SyncUi
 * Classes for informing UI components to sync/refresh their display
 */

/**
 * @docs-group-description WorkflowTask
 * Classes for working a Workflow or Task
 */
/**
 * @docs-group-description Tools
 * Core Tool definitions
 */
/**
 * @docs-group-description ToolSettings
 * Classes for working Tool Settings
 */
/**
 * @docs-group-description Utilities
 * Various utility classes for working with a UI and Redux.
 */
/**
 * @docs-group-description Widget
 * Classes for working with a Widget
 */
/**
 * @docs-group-description Zone
 * Classes for working with a Zone
 */
