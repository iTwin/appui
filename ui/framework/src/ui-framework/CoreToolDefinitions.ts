/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tools */

// cSpell:ignore configurableui
import {
  FitViewTool, FlyViewTool, IModelApp, PanViewTool, RotateViewTool, SelectionTool, ViewToggleCameraTool, WalkViewTool,
  WindowAreaTool, ZoomViewTool, ViewClipByPlaneTool,
  ViewClipDecorationProvider,
} from "@bentley/imodeljs-frontend";
import { ViewFlags } from "@bentley/imodeljs-common";
import { ToolItemDef } from "./shared/ToolItemDef";

/** Utility Class that provides definitions of tools provided by iModel.js core. These definitions can be used to populate the UI.
 * @public
 */
// istanbul ignore next
export class CoreTools {
  public static get fitViewCommand() {
    return new ToolItemDef({
      toolId: FitViewTool.toolId,
      iconSpec: "icon-fit-to-view",
      label: () => FitViewTool.flyover,
      tooltip: () => FitViewTool.description,
      execute: () => { IModelApp.tools.run(FitViewTool.toolId, IModelApp.viewManager.selectedView, true); },
    });
  }

  public static get windowAreaCommand() {
    return new ToolItemDef({
      toolId: WindowAreaTool.toolId,
      iconSpec: "icon-window-area",
      label: () => WindowAreaTool.flyover,
      tooltip: () => WindowAreaTool.description,
      execute: () => { IModelApp.tools.run(WindowAreaTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get zoomViewCommand() {
    return new ToolItemDef({
      toolId: ZoomViewTool.toolId,
      iconSpec: "icon-zoom",
      label: () => ZoomViewTool.flyover,
      tooltip: () => ZoomViewTool.description,
      execute: () => { IModelApp.tools.run(ZoomViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get panViewCommand() {
    return new ToolItemDef({
      toolId: PanViewTool.toolId,
      iconSpec: "icon-hand-2",
      label: () => PanViewTool.flyover,
      tooltip: () => PanViewTool.description,
      execute: () => { IModelApp.tools.run(PanViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get rotateViewCommand() {
    return new ToolItemDef({
      toolId: RotateViewTool.toolId,
      iconSpec: "icon-gyroscope",
      label: () => RotateViewTool.flyover,
      tooltip: () => RotateViewTool.description,
      execute: () => { IModelApp.tools.run(RotateViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get walkViewCommand() {
    return new ToolItemDef({
      toolId: WalkViewTool.toolId,
      iconSpec: "icon-walk",
      label: () => WalkViewTool.flyover,
      tooltip: () => WalkViewTool.description,
      execute: () => { IModelApp.tools.run(WalkViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get selectElementCommand() {
    return new ToolItemDef({
      toolId: SelectionTool.toolId,
      iconSpec: "icon-cursor",
      label: () => SelectionTool.flyover,
      tooltip: () => SelectionTool.description,
      execute: () => { IModelApp.tools.run(SelectionTool.toolId); },
    });
  }

  public static get toggleCameraViewCommand() {
    return new ToolItemDef({
      toolId: ViewToggleCameraTool.toolId,
      iconSpec: "icon-camera",
      label: () => ViewToggleCameraTool.flyover,
      tooltip: () => ViewToggleCameraTool.description,
      execute: () => { IModelApp.tools.run(ViewToggleCameraTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get flyViewCommand() {
    return new ToolItemDef({
      toolId: FlyViewTool.toolId,
      iconSpec: "icon-airplane",
      label: () => FlyViewTool.flyover,
      tooltip: () => FlyViewTool.description,
      execute: () => { IModelApp.tools.run(FlyViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  // note current ViewClipByPlaneTool is not automatically registered so the app must call ViewClipByPlaneTool.register();
  public static get sectionByPlaneCommand() {
    return new ToolItemDef({
      toolId: ViewClipByPlaneTool.toolId,
      iconSpec: "icon-plane",
      label: () => ViewClipByPlaneTool.flyover,
      tooltip: () => ViewClipByPlaneTool.description,
      execute: () => {
        const vp = IModelApp.viewManager.selectedView;
        if (!vp || !vp.view.is3d())
          return;

        // Turn on clip volume flag for section tools
        const viewFlags: ViewFlags = vp.view.viewFlags.clone();
        viewFlags.clipVolume = true;
        vp.viewFlags = viewFlags;

        IModelApp.tools.run(ViewClipByPlaneTool.toolId, ViewClipDecorationProvider.create());
      },
    });
  }

}
