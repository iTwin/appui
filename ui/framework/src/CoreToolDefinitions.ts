/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tools */

import {
  IModelApp, SelectionTool, FitViewTool, WindowAreaTool,
  PanViewTool, RotateViewTool, ViewToggleCameraTool, WalkViewTool, FlyViewTool, ZoomViewTool,
} from "@bentley/imodeljs-frontend";

// cSpell:ignore configurableui
import { ToolItemDef } from "./configurableui/Item";
import { ContentViewManager } from "./configurableui/ContentViewManager";
import { BaseItemState } from "./configurableui/ItemDefBase";
import { SyncUiEventId } from "./SyncUiEventDispatcher";
import { AnalysisAnimationTool } from "./tools/AnalysisAnimation";

/** Utility Class that provides definitions of tools provided by imodel.js core. These definitions can be used to populate the ui. */
export class CoreTools {
  public static get fitViewCommand() {
    return new ToolItemDef({
      toolId: FitViewTool.toolId,
      iconSpec: "icon-fit-to-view",
      labelKey: "CoreTools:tools.View.Fit.flyover",
      tooltipKey: "CoreTools:tools.View.Fit.description",
      execute: () => { IModelApp.tools.run(FitViewTool.toolId, IModelApp.viewManager.selectedView, true); },
    });
  }

  public static get windowAreaCommand() {
    return new ToolItemDef({
      toolId: WindowAreaTool.toolId,
      iconSpec: "icon-window-area",
      labelKey: "CoreTools:tools.View.WindowArea.flyover",
      tooltipKey: "CoreTools:tools.View.WindowArea.description",
      execute: () => { IModelApp.tools.run(WindowAreaTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get zoomViewCommand() {
    return new ToolItemDef({
      toolId: ZoomViewTool.toolId,
      iconSpec: "icon-zoom",
      labelKey: "CoreTools:tools.View.Zoom.flyover",
      tooltipKey: "CoreTools:tools.View.Zoom.description",
      execute: () => { IModelApp.tools.run(ZoomViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get panViewCommand() {
    return new ToolItemDef({
      toolId: PanViewTool.toolId,
      iconSpec: "icon-hand-2",
      labelKey: "CoreTools:tools.View.Pan.flyover",
      tooltipKey: "CoreTools:tools.View.Pan.description",
      execute: () => { IModelApp.tools.run(PanViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get rotateViewCommand() {
    return new ToolItemDef({
      toolId: RotateViewTool.toolId,
      iconSpec: "icon-rotate-left",
      labelKey: "CoreTools:tools.View.Rotate.flyover",
      tooltipKey: "CoreTools:tools.View.Rotate.description",
      execute: () => { IModelApp.tools.run(RotateViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get walkViewCommand() {
    return new ToolItemDef({
      toolId: WalkViewTool.toolId,
      iconSpec: "icon-walk",
      labelKey: "CoreTools:tools.View.Walk.flyover",
      tooltipKey: "CoreTools:tools.View.Walk.description",
      execute: () => { IModelApp.tools.run(WalkViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get selectElementCommand() {
    return new ToolItemDef({
      toolId: SelectionTool.toolId,
      iconSpec: "icon-cursor",
      labelKey: "CoreTools:tools.Select.flyover",
      tooltipKey: "CoreTools:tools.Select.description",
      execute: () => { IModelApp.tools.run(SelectionTool.toolId); },
    });
  }

  public static get toggleCameraViewCommand() {
    return new ToolItemDef({
      toolId: ViewToggleCameraTool.toolId,
      iconSpec: "icon-camera",
      labelKey: "UiFramework:tools.View.ToggleCamera.flyover",
      tooltipKey: "UiFramework:tools.View.ToggleCamera.description",
      execute: () => { IModelApp.tools.run(ViewToggleCameraTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get flyViewCommand() {
    return new ToolItemDef({
      toolId: FlyViewTool.toolId,
      iconSpec: "icon-airplane",
      labelKey: "UiFramework:tools.View.Fly.flyover",
      tooltipKey: "UiFramework:tools.View.Fly.description",
      execute: () => { IModelApp.tools.run(FlyViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get analysisAnimationCommand() {
    return new ToolItemDef({
      toolId: AnalysisAnimationTool.toolId,
      iconSpec: "icon-camera-animation",
      labelKey: "UiFramework:tools.AnalysisAnimation.flyover",
      tooltipKey: "UiFramework:tools.AnalysisAnimation.description",
      execute: () => { IModelApp.tools.run(AnalysisAnimationTool.toolId); },
      stateSyncIds: [SyncUiEventId.ActiveContentChanged],
      stateFunc: (currentState: Readonly<BaseItemState>): BaseItemState => {
        const returnState: BaseItemState = { ...currentState };
        const activeContentControl = ContentViewManager.getActiveContentControl();

        if (activeContentControl && activeContentControl.viewport && (undefined !== activeContentControl.viewport.view.AnalysisStyle))
          returnState.isVisible = true;
        else
          returnState.isVisible = false;
        return returnState;
      },
    });
  }
}
