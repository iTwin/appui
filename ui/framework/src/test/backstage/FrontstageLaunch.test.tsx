/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { mount, shallow } from "enzyme";
import { expect } from "chai";
import * as sinon from "sinon";
import {
  FrontstageLaunchBackstageItem,
  FrontstageManager,
  FrontstageActivatedEventArgs,
  ConfigurableUiManager,
  FrontstageProvider,
  Frontstage,
  FrontstageProps,
  BackstageItemState,
} from "../../ui-framework";
import TestUtils from "../TestUtils";
import { BackstageItem as NZ_BackstageItem } from "@bentley/ui-ninezone";
import { CoreTools } from "../../ui-framework/CoreToolDefinitions";
import { SyncUiEventDispatcher } from "../../ui-framework/syncui/SyncUiEventDispatcher";
import { Logger } from "@bentley/bentleyjs-core";

describe("Backstage", () => {
  const testEventId = "test-state-function-event";

  before(async () => {
    await TestUtils.initializeUiFramework();

    FrontstageManager.setActiveFrontstageDef(undefined); // tslint:disable-line:no-floating-promises
  });

  after(() => {
    TestUtils.terminateUiFramework();
  });

  describe("<FrontstageLaunchBackstageItem />", () => {
    it("FrontstageLaunchBackstageItem should render & execute", async () => {
      const spyMethod = sinon.stub();
      let stateFuncRun = false;
      const stateFunc = (state: Readonly<BackstageItemState>): BackstageItemState => {
        stateFuncRun = true;
        return { ...state, isActive: true } as BackstageItemState;
      };

      class Frontstage1 extends FrontstageProvider {
        public get frontstage(): React.ReactElement<FrontstageProps> {
          return (
            <Frontstage
              id="Test1"
              defaultTool={CoreTools.selectElementCommand}
              defaultLayout="FourQuadrants"
              contentGroup="TestContentGroup1"
            />
          );
        }
      }
      ConfigurableUiManager.addFrontstageProvider(new Frontstage1());

      const remove = FrontstageManager.onFrontstageActivatedEvent.addListener((_args: FrontstageActivatedEventArgs) => spyMethod());
      const wrapper = mount(
        <FrontstageLaunchBackstageItem frontstageId="Test1" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder"
          isEnabled={true} isActive={false}
          stateSyncIds={[testEventId]} stateFunc={stateFunc} />,
      );

      expect(stateFuncRun).to.be.false;
      SyncUiEventDispatcher.dispatchImmediateSyncUiEvent(testEventId);
      expect(stateFuncRun).to.be.true;
      wrapper.update();

      const backstageItem = wrapper.find(NZ_BackstageItem);
      backstageItem.find(".nz-backstage-item").simulate("click");

      await TestUtils.flushAsyncOperations();
      expect(spyMethod.calledOnce).to.be.true;
      remove();
      wrapper.unmount();
    });

    it("FrontstageLaunchBackstageItem should log error when invalid frontstageId is provided", async () => {
      const spyMethod = sinon.spy(Logger, "logError");
      const wrapper = mount(
        <FrontstageLaunchBackstageItem frontstageId="BadTest" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" />,
      );

      const backstageItem = wrapper.find(NZ_BackstageItem);
      backstageItem.find(".nz-backstage-item").simulate("click");
      spyMethod.calledOnce.should.true;

      wrapper.unmount();
      (Logger.logError as any).restore();
    });

    it("FrontstageLaunchBackstageItem renders correctly when inactive", async () => {
      await FrontstageManager.setActiveFrontstageDef(undefined);
      const wrapper = shallow(<FrontstageLaunchBackstageItem frontstageId="Test1" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" />);
      wrapper.should.matchSnapshot();
      wrapper.unmount();
    });

    it("FrontstageLaunchBackstageItem renders correctly when active", async () => {
      const frontstageDef = FrontstageManager.findFrontstageDef("Test1");
      expect(frontstageDef).to.not.be.undefined;

      if (frontstageDef) {
        await FrontstageManager.setActiveFrontstageDef(frontstageDef);
        const wrapper = shallow(<FrontstageLaunchBackstageItem frontstageId="Test1" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" />);
        wrapper.should.matchSnapshot();
        wrapper.unmount();
      }
    });

    it("FrontstageLaunchBackstageItem updates on frontstage activation", async () => {
      await FrontstageManager.setActiveFrontstageDef(undefined);
      const wrapper = mount(<FrontstageLaunchBackstageItem frontstageId="Test1" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" />);
      expect(wrapper.find("li.nz-active").length).to.eq(0);

      const frontstageDef = FrontstageManager.findFrontstageDef("Test1");
      expect(frontstageDef).to.not.be.undefined;

      if (frontstageDef) {
        await FrontstageManager.setActiveFrontstageDef(frontstageDef);
        wrapper.update();
        expect(wrapper.find("li.nz-active").length).to.eq(1);
      }

      wrapper.unmount();
    });

    it("FrontstageLaunchBackstageItem updates on property change", async () => {
      const wrapper = mount(<FrontstageLaunchBackstageItem frontstageId="Test1" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" isEnabled={false} />);
      expect(wrapper.find("li.nz-disabled").length).to.eq(1);

      wrapper.setProps({ isEnabled: true });
      wrapper.update();
      expect(wrapper.find("li.nz-disabled").length).to.eq(0);

      wrapper.unmount();
    });
  });
});
