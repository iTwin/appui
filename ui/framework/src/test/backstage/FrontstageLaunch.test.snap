// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backstage <FrontstageLaunchBackstageItem /> FrontstageLaunchBackstageItem renders correctly when active 1`] = `
<Component
  icon={
    <Unknown
      iconSpec="icon-placeholder"
    />
  }
  isActive={true}
  isDisabled={false}
  key="Test1"
  onClick={[Function]}
  subtitle=""
>
  tests.label
</Component>
`;

exports[`Backstage <FrontstageLaunchBackstageItem /> FrontstageLaunchBackstageItem renders correctly when inactive 1`] = `
<Component
  icon={
    <Unknown
      iconSpec="icon-placeholder"
    />
  }
  isActive={false}
  isDisabled={false}
  key="Test1"
  onClick={[Function]}
  subtitle=""
>
  tests.label
</Component>
`;
